src/engine/TPA.cc

```diff
@@ -1485,186 +1457,145 @@ VerificationResult TPAEngine::solveTransitionSystemGraph(const ChcDirectedGraph
 void TransitionSystemNetworkManager::initNetwork() {
     assert(networkMap.empty());
     for (auto vid : graph.getVertices()) {
-        if (vid == graph.getEntry() or vid == graph.getExit()) {
-            networkMap.insert({vid, NetworkNode()});
-            continue;
-        }
-        networkMap.insert({vid, NetworkNode()});
-        auto & node = networkMap.at(vid);
-        node.solver = mkSolver();
+        auto [it, _] = networkMap.insert({vid, NetworkNode()});
+        auto & node = it->second;
+        node.preSafe = logic.getTerm_false();
+        node.postSafe = logic.getTerm_false();
         node.blocked_children = 0;
+        if (vid == graph.getEntry() or vid == graph.getExit()) { continue; }
+        node.solver = mkSolver();
         TransitionSystem ts = constructTransitionSystemFor(vid);
         node.solver->resetTransitionSystem(ts);
-        node.trulySafe = logic.getTerm_false();
-    }
-    TimeMachine tm(logic);
-    for (EId eid : adjacencyRepresentation.getOutgoingEdgesFor(graph.getEntry())) {
-        auto target = graph.getTarget(eid);
-        getNode(target).solver->resetInitialStates(logic.getTerm_true());
-    }
-    for (EId eid : adjacencyRepresentation.getIncomingEdgesFor(graph.getExit())) {
-        auto source = graph.getSource(eid);
-        networkMap.at(source).solver->updateQueryStates(logic.getTerm_true());
     }
     // Connect the network
+    auto isSelfLoop = [this](EId eid) { return graph.getSource(eid) == graph.getTarget(eid); };
+    // TODO: no need for reverse post order, better to process outgoing edges
```

Comment:
I am not sure if it can be systematically analyzed...
Causation is really fluent in here, I'm not sure we would be able to come up with specific way to pick edges to traverse before we actually traverse them...

Reply:
That's true. I would just like to avoid computing the reverse post order, if possible.
If it does not hurt performance, then this code can be simplified.